pmd-java/src/main/resources/category/java/codestyle.xml

```diff
@@ -1353,6 +1353,55 @@ public class ClassInDefaultPackage {
         </example>
     </rule>
 
+    <rule name="UseUnderscoresInNumericLiterals"
+          language="java"
+          since="6.10.0"
+          minimumLanguageVersion="1.7"
+          message="Numeric literals with more than 3 digits must use '_' as a separator"
```

Comment:
Maybe the rule message can be improved by using the placeholder `{0}`, which will be replaced by the image of the matched node, i.e. by the number. E.g. `Number {0} should separate every third digit with an underscore`. This message also passes when the violation is because of misplaced underscores

Reply:
This works fine! Could you tell me how {0} is associated to Image?

Reply:
Java rules may report a violation with an array of arguments, which are interpolated index-by-index on placeholder like `{0}`, `{1}`, etc. As it happens XPath rules have no flexible way to add arguments to a message, but the image of the matched node is always passed nevertheless, so is always accessible as `{0}` (for XPath rules only). See here:

https://github.com/pmd/pmd/blob/0c1afc426debff979b25219dbc8f4e07c76645b5/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/XPathRule.java#L114